src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala

```diff
@@ -97,28 +99,50 @@ trait WorkspaceComponent {
-          val deleteActions = deleteWorkspaceAttributes(wsIdAndAttributeRecords.map(_._2))
+      uniqueResult[WorkspaceRecord](findByIdQuery(UUID.fromString(workspace.workspaceId))) flatMap {
+        case None =>
+          (workspaceQuery += marshalWorkspace(workspace)) andThen
+            insertOrUpdateAccessRecords(workspace) andThen
+            insertAttributes(workspace)
+        case Some(workspaceRecord) =>
+          insertOrUpdateAccessRecords(workspace) andThen
+            deleteAttributes(workspace) andThen
+            insertAttributes(workspace) andThen
+            optimisticLockUpdate(workspaceRecord)
+      } map { _ => workspace }
+    }
+
+    private def insertOrUpdateAccessRecords(workspace: Workspace): WriteAction[Unit] = {
+      val id = UUID.fromString(workspace.workspaceId)
+      val accessRecords = workspace.accessLevels.map { case (accessLevel, group) => WorkspaceAccessRecord(id, group.groupName.value, accessLevel.toString, false) }
+      val realmAclRecords = workspace.realmACLs.map { case (accessLevel, group) => WorkspaceAccessRecord(id, group.groupName.value, accessLevel.toString, true) }
+      DBIO.seq((accessRecords ++ realmAclRecords).map { workspaceAccessQuery insertOrUpdate }.toSeq: _*)
+    }
+
+    private def deleteAttributes(workspace: Workspace): ReadWriteAction[Unit] = {
+      val workspaceId = UUID.fromString(workspace.workspaceId)
+      workspaceAttributes(findByIdQuery(workspaceId)).result.flatMap { wsIdAndAttributeRecords =>
+        // clear existing attributes, any concurrency locking should be handled at the workspace level
+        val deleteActions = deleteWorkspaceAttributes(wsIdAndAttributeRecords.map(_._2))
+        DBIO.seq(deleteActions.toSeq:_*)
+      }
+    }
 
-          // insert attributes
-          val insertActions = workspace.attributes.flatMap { case (name, attribute) =>
-            attributeQuery.insertAttributeRecords(name, attribute, workspaceRecord.id).map(_.flatMap{ attrId =>
-              insertWorkspaceAttributeMapping(workspaceRecord, attrId)
-            })
-          }
+    private def insertAttributes(workspace: Workspace): ReadWriteAction[Unit] = {
+      val workspaceId = UUID.fromString(workspace.workspaceId)
+      val insertActions = workspace.attributes.flatMap { case (name, attribute) =>
+        attributeQuery.insertAttributeRecords(name, attribute, workspaceId).map(_.flatMap{ attrId =>
+          insertWorkspaceAttributeMapping(workspaceId, attrId)
+        })
+      }
+      DBIO.seq(insertActions.toSeq:_*)
+    }
 
-          DBIO.seq(deleteActions ++ insertActions:_*)
-        }
-      } map(_ => workspace)
+    private def optimisticLockUpdate(originalRec: WorkspaceRecord): ReadWriteAction[Int] = {
+      findByIdAndRecordVersionQuery(originalRec.id, originalRec.recordVersion) update originalRec.copy(recordVersion = originalRec.recordVersion + 1) map {
```

Comment:
Look for the current version and update by 1.  If the current version is not present, someone else is currently updating this: throw an exception and abort the transaction.